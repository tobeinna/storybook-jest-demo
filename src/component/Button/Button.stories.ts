import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import Button from "./Button";

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: "Components/Button",
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    type: {
      control: 'select',
      description: "Define the button's type",
    },
    text: {
      control: "text",
      description: "Define the button's displayed label",
    },
    disabled: {
      control: "boolean",
      description: "Is the button disabled?",
    },
    onClick: {
      description: "Called when the button's onClick event triggered",
    },
  },
  // Use `fn` to spy on the onClick arg, which will appear in the actions panel once invoked: https://storybook.js.org/docs/essentials/actions#action-args
  args: { onClick: fn() },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const Primary: Story = {
  args: {
    type: "primary",
    text: "Primary Button",
    disabled: false,
  },
};

export const Secondary: Story = {
  args: {
    type: "secondary",
    text: "Secondary Button",
    disabled: false,
  },
};

export const Danger: Story = {
  args: {
    type: "danger",
    text: "Danger Button",
    disabled: false,
  },
};

export const Disabled: Story = {
  args: {
    type: "primary",
    text: "Disabled Primary Button",
    disabled: true,
  },
};
