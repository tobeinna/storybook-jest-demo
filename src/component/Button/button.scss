button {
  padding: 12px 24px;
  opacity: 1;
  border-radius: 6px;
  span {
    color: #e2e0e0;
  }
}

button:hover {
  transition: background 0.25s;
  span {
    color: #ffffff;
    transition: color 0.25s;
  }
}

button:disabled {
  opacity: 0.7;
}

.primary {
  background-color: #4197f9;
}

.primary:hover {
  background-color: #4c9ffe;
}

.secondary {
  background-color: #434343;
}

.secondary:hover {
  background-color: #505050;
}

.danger {
  border: 1px solid #ff4242;
  span {
    color: #ff4242;
  }
}

.danger:hover {
  border: 1px solid #fd6464 !important;
  span {
    color: #fd6464;
  }
}
