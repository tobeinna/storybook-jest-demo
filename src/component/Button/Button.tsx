import './button.scss';

interface ButtonProps {
    text?: string,
    type: 'primary' | 'secondary' | 'danger',
    disabled?: boolean,
    onClick?: () => void,
}

function Button({ text, type = 'primary', disabled = false, onClick }: ButtonProps) {
    return <button className={type} disabled={disabled} onClick={onClick}>
        <span>{text}</span>
    </button>;
}

export default Button;