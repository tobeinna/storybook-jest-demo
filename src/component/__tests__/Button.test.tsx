import { render, screen, fireEvent } from '@testing-library/react';
import Button from '../Button/Button';

describe('Button component', () => {
    test('snapshot test for primary button', () => {
        const tree = render(<Button text="Primary Button" type="primary" />).container;
        expect(tree).toMatchSnapshot();
    });

    test('snapshot test for secondary button', () => {
        const tree = render(<Button text="Secondary Button" type="secondary" />).container;
        expect(tree).toMatchSnapshot();
    });

    test('snapshot test for danger button', () => {
        const tree = render(<Button text="Danger Button" type="danger" />).container;
        expect(tree).toMatchSnapshot();
    });

    test('snapshot test for disabled button', () => {
        const tree = render(<Button text="Disabled Button" type="primary" disabled />).container;
        expect(tree).toMatchSnapshot();
    });

    test('render with correct text', () => {
        render(<Button text="Button" type="primary" />);
        const element = screen.getByText(/Button/i);
        expect(element).toBeInTheDocument();
    });

    test('check the class when rerender by changing type', () => {
        const { rerender } = render(<Button text="Primary Button" type="primary" />);
        const element = screen.getByText(/Primary Button/i);
        expect(element).toHaveClass('primary');

        rerender(<Button text="Danger Button" type="danger" />);
        expect(element).toHaveClass('danger');
    });

    test('disable the button when disabled prop is true', () => {
        render(<Button text="Button" type="primary" disabled />);
        const element = screen.getByText(/Button/i);
        expect(element).toBeDisabled();
    });

    test('call onClick when clicked', () => {
        const handleClick = jest.fn();
        render(<Button text="Button" type="primary" onClick={handleClick} />);
        const element = screen.getByText(/Button/i);
        fireEvent.click(element);
        expect(handleClick).toHaveBeenCalled();
    });

    test('not call onClick when button is disabled', () => {
        const handleClick = jest.fn();
        render(<Button text="Button" type="primary" disabled onClick={handleClick} />);
        const element = screen.getByText(/Button/i);
        fireEvent.click(element);
        expect(handleClick).not.toHaveBeenCalled();
    });
});
