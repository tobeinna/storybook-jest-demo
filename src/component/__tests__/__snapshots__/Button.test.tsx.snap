// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component snapshot test for danger button 1`] = `
<div>
  <button
    class="danger"
  >
    <span>
      Danger Button
    </span>
  </button>
</div>
`;

exports[`Button component snapshot test for disabled button 1`] = `
<div>
  <button
    class="primary"
    disabled=""
  >
    <span>
      Disabled Button
    </span>
  </button>
</div>
`;

exports[`Button component snapshot test for primary button 1`] = `
<div>
  <button
    class="primary"
  >
    <span>
      Primary Button
    </span>
  </button>
</div>
`;

exports[`Button component snapshot test for secondary button 1`] = `
<div>
  <button
    class="secondary"
  >
    <span>
      Secondary Button
    </span>
  </button>
</div>
`;
